File: day19/day19.kt
import java.io.File

fun main() {
    val input = File("day19.txt").readText()
    val (allRules, allMessages) = input.split("\n\n")
    val rules = allRules.split("\n").map { it.split(": ") }.map {
        (k, v) -> k.toInt() to if (v[0] == '"') v[1].toString() else
            v.split(" | ").map { it.split(" ").map { it.toInt() } }
    }.toMap()
    val messages = allMessages.trimEnd().split("\n")

    // Part 1
    val options = mutableMapOf<Int,Set<String>>()
    rules.forEach { k, v -> if (v is String) options[k] = setOf(v) }

    fun build(key: Int): Set<String> {
        return options.getOrElse(key) {
            val rule = rules[key] as List<List<Int>>
            rule.flatMap {
                it.fold(listOf("")) {
                    a, b -> a.flatMap { s -> build(b).map { s + it } }
                }
            }.toSet().also { options[key] = it }
        }
    }

    build(0)
    println(messages.count { it in options[0]!! })

    // Part 2
    val l42 = options[42]!!.first().length
    val l31 = options[31]!!.first().length
    val matches_8 = { msg: String ->
        options[42]!!.any { msg.startsWith(it) } &&
        (msg.length == l42 || matches_8(msg.drop(l42))) }
    val matches_11 = { msg: String ->
        options[42]!!.any { msg.startsWith(it) } &&
        options[31]!!.any { msg.endsWith(it) } &&
        (msg.length == l42 + l31 || matches_11(msg.drop(l42).dropLast(l31))) }
    val matches_0 = { msg: String -> (0..msg.length).any {
        matches_8(msg.take(it)) && matches_11(msg.drop(it)) } }
    println(messages.count { matches_0(it) })
}

File: day19/day19.py
input_ = open("day19.txt").read()
all_rules, all_messages = input_.split('\n\n')
rules = {int(k): v[1] if v[0] == '"' else
         [tuple(map(int, r.split(' '))) for r in v.split(' | ')]
         for k, v in [s.split(': ') for s in all_rules.split('\n')]}
messages = all_messages.rstrip().split('\n')

# Part 1
options = {k: {v} for k, v in rules.items() if isinstance(v, str)}
def build(key):
    if key in options:
        return options[key]
    values = set()
    for seq in rules[key]:
        left = ['']
        for right in map(build, seq):
            left = [l + r for l in left for r in right]
        values.update(left)
    options[key] = values
    return values

build(0)
print(sum(s in options[0] for s in messages))

# Part 2
l42 = len(next(iter(options[42])))
l31 = len(next(iter(options[31])))

def matches_8(msg):
    return any(msg.startswith(p) for p in options[42]) and \
        (len(msg) == l42 or matches_8(msg[l42:]))

def matches_11(msg):
    return any(msg.startswith(p) for p in options[42]) and \
        any(msg.endswith(s) for s in options[31]) and \
        (len(msg) == l42 + l31 or matches_11(msg[l42:-l31]))

def matches_0(msg):
    return any(matches_8(msg[:i]) and matches_11(msg[i:])
               for i in range(0, len(msg)))

print(sum(matches_0(s) for s in messages))
